src/projection.jl

```diff
@@ -0,0 +1,140 @@
+struct ProjectTo{P, D<:NamedTuple}
+    info::D
+end
+ProjectTo{P}(info::D) where {P,D<:NamedTuple} = ProjectTo{P,D}(info)
+ProjectTo{P}(; kwargs...) where {P} = ProjectTo{P}(NamedTuple(kwargs))
+
+backing(project::ProjectTo) = getfield(project, :info)
+Base.getproperty(p::ProjectTo, name::Symbol) = getproperty(backing(p), name)
+Base.propertynames(p::ProjectTo) = propertynames(backing(p))
+
+function Base.show(io::IO, project::ProjectTo{T}) where T
+    print(io, "ProjectTo{")
+    show(io, T)
+    print(io, "}")
+    if isempty(backing(project))
+        print(io, "()")
+    else
+        show(io, backing(project))
+    end
+end
+
+
+"""
+    ProjectTo(x)
+
+Returns a `ProjectTo{P,...}` functor able to project a differential `dx` onto the type `T`
+for a primal `x`.
+This functor encloses over what ever is needed to be able to be able to do that projection.
+For example, when projecting `dx=ZeroTangent()` on an array `P=Array{T, N}`, the size of `x`
+is not available from `P`, so it is stored in the functor.
+"""
+function ProjectTo end
+
+"""
+    (::ProjectTo{T})(dx)
+
+Projects the differential `dx` on the onto type `T`.
+`ProjectTo{T}` is a functor that knows how to perform this projection.
+"""
+function (::ProjectTo) end
```

Comment:
Apparently this doensn't work. https://github.com/JuliaDocs/Documenter.jl/issues/228
Which explains the warning

I guess Append theing one the the constructor docstring and hang it off the struct.